File: aichallenge/workspace/src/aichallenge_submit/dwa/dwa/dwa/simulator.py
import math

class Simulator_DWA_robot:
    def __init__(self):
        # 加速度制限
        self.max_accelation = 1.0
        self.max_ang_accelation = 100 * math.pi / 180
        # 速度制限
        self.lim_max_velo = 1.6  # m/s
        self.lim_min_velo = 0.0  # m/s
        self.lim_max_ang_velo = math.pi
        self.lim_min_ang_velo = -math.pi

    def predict_state(self, ang_velo, velo, x, y, th, dt, pre_step):
        next_xs = []
        next_ys = []
        next_ths = []

        for _ in range(pre_step):
            temp_x = velo * math.cos(th) * dt + x
            temp_y = velo * math.sin(th) * dt + y
            temp_th = ang_velo * dt + th

            next_xs.append(temp_x)
            next_ys.append(temp_y)
            next_ths.append(temp_th)

            x = temp_x
            y = temp_y
            th = temp_th

        return next_xs, next_ys, next_ths

File: aichallenge/workspace/src/aichallenge_submit/dwa/simple_dwa/src/simulator.cpp
#include "simulator.hpp"
#include <cmath>
#include <rclcpp/rclcpp.hpp>
Simulator::Simulator() {}

Simulator::Simulator(const Parameters &params)
    : params_(params) {}

std::vector<double> Simulator::predictState(double ang_velo, double velo, double x, double y, double th, double dt, int pre_step)
{
  std::vector<double> next_xs;
  std::vector<double> next_ys;
  std::vector<double> next_ths;
  for (int i = 0; i < pre_step; ++i)
  {
    double temp_x = velo * std::cos(th) * dt + x;
    double temp_y = velo * std::sin(th) * dt + y;
    double temp_th = ang_velo * dt + th;

    next_xs.push_back(temp_x);
    next_ys.push_back(temp_y);
    next_ths.push_back(temp_th);

    x = temp_x;
    y = temp_y;
    th = temp_th;
  }

  std::vector<double> result;
  result.insert(result.end(), next_xs.begin(), next_xs.end());
  result.insert(result.end(), next_ys.begin(), next_ys.end());
  result.insert(result.end(), next_ths.begin(), next_ths.end());
  return result;
}